api/agent/pure_runner.go

```diff
@@ -0,0 +1,491 @@
+package agent
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	runner "github.com/fnproject/fn/api/agent/grpc"
+	"github.com/fnproject/fn/api/models"
+	"github.com/go-openapi/strfmt"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+)
+
+// callHandle represents the state of the call as handled by the pure runner, and additionally it implements the
+// interface of http.ResponseWriter so that it can be used for streaming the output back.
+type callHandle struct {
+	engagement runner.RunnerProtocol_EngageServer
+	c          *call // the agent's version of call
+	input      io.WriteCloser
+	started    bool
+	// As the state can be set and checked by both goroutines handling this state, we need a mutex.
+	stateMutex sync.Mutex
```

Comment:
There are many network calls while holding this mutex locked. Is that intended? If yes, then we do have upper bounds on timeouts on all of those rpc calls?

Reply:
Each call (as in, the call of an Fn function) has a different handle and therefore a different mutex - it's not a sequential chokepoint.

Within a single Fn function call, there are only two goroutines that access the shared state: the receive loop and the code that runs Submit and sends back the output. The receive loop is mostly idle once the function has received the input, so there shouldn't be much contention on the mutex.